### PR TITLE
Mt7921e devel - MTK internal SQC fixes

### PR DESCRIPTION
**Pull request by objelf**

### CODE REVIEW

File: drivers/net/wireless/mediatek/mt76/mt7921/regs.h [hunk 1]
```diff
@@ -36,18 +36,6 @@
...
-#define MT_MDP_RCFR0_MCU_RX_CTL_NON_BAR	GENMASK(7, 6)
-#define MT_MDP_RCFR0_MCU_RX_CTL_BAR	GENMASK(9, 8)
-
-#define MT_MDP_BNRCFR1(_band)		MT_MDP(0x074 + ((_band) << 8))
-#define MT_MDP_RCFR1_MCU_RX_BYPASS	GENMASK(23, 22)
-#define MT_MDP_RCFR1_RX_DROPPED_UCAST	GENMASK(28, 27)
-#define MT_MDP_RCFR1_RX_DROPPED_MCAST	GENMASK(30, 29)
-#define MT_MDP_TO_HIF			0
-#define MT_MDP_TO_WM			1
-
```
Comment (LorenzoBianconi) [inline drivers/net/wireless/mediatek/mt76/mt7921/regs.h]: @objelf I would not drop these definitions even if they are not used at the moment, maybe we need them in the future

File: drivers/net/wireless/mediatek/mt76/mt76_connac.h [hunk 2]
```diff
@@ -134,12 +134,16 @@ mt76_connac_pm_unref(struct mt76_connac_pm *pm)
...
+	struct mt76_dev *dev = phy->dev;
 	bool ret;
 
 	spin_lock_bh(&pm->wake.lock);
 	ret = pm->wake.count || test_and_set_bit(MT76_STATE_PM, &phy->state);
 	spin_unlock_bh(&pm->wake.lock);
 
+	if (mt76_is_mmio(dev) && dev->token_count)
+		return true;
+
```
Comment (LorenzoBianconi) [inline drivers/net/wireless/mediatek/mt76/mt76_connac.h]: @objelf I guess you need to check token_count before setting MT76_STATE_PM otherwise we will return true (so not performing fw_own) and state is marked in PM state. Another approach would be to clear MT76_STATE_PM before returning true

File: drivers/net/wireless/mediatek/mt76/mt7921/mcu.c [hunk 3]
```diff
@@ -981,6 +981,7 @@ int mt7921_run_firmware(struct mt7921_dev *dev)
 
 	set_bit(MT76_STATE_MCU_RUNNING, &dev->mphy.state);
 	mt7921_mcu_fw_log_2_host(dev, 1);
+	mt76_connac_mcu_set_deep_sleep(&dev->mt76, dev->pm.enable);
```
Comment (LorenzoBianconi) [inline drivers/net/wireless/mediatek/mt76/mt7921/mcu.c]: @objelf this is not correct since after mt7921_mcu_init() there are other device accesses without grabbing dev mutex, I guess we need to move it at the end of mt7921_register_device

File: drivers/net/wireless/mediatek/mt76/mt76_connac_mcu.h [hunk 4]
```diff
@@ -1090,5 +1090,7 @@ int mt76_connac_mcu_set_roc(struct mt76_dev *dev, struct ieee80211_vif *vif,
 void mt76_connac_mcu_roc_event(struct mt76_phy *phy,
 			       struct mt76_connac_roc *roc,
 			       struct mt76_connac_roc_tlv *event);
-
+int mt76_connac_sta_state_dp(struct mt76_dev *dev,
```
Comment (LorenzoBianconi) [inline drivers/net/wireless/mediatek/mt76/mt76_connac_mcu.h:1114]: @objelf I think this will not apply upstream :)

Comment (objelf) [reply drivers/net/wireless/mediatek/mt76/mt76_connac_mcu.h:1114]: do you mean it is not neccessary to remove the blank line, right ?

File: drivers/net/wireless/mediatek/mt76/mt7921/mcu.c (hunk 3 above)
Comment (objelf) [reply drivers/net/wireless/mediatek/mt76/mt7921/mcu.c]: ack, i will check about it.

File: drivers/net/wireless/mediatek/mt76/mt7921/regs.h (hunk 1 above)
Comment (objelf) [reply drivers/net/wireless/mediatek/mt76/mt7921/regs.h]: ok, lets keep it.

File: drivers/net/wireless/mediatek/mt76/mt76_connac.h (hunk 2 above)
Comment (objelf) [reply drivers/net/wireless/mediatek/mt76/mt76_connac.h]: ack i will check about it.

File: drivers/net/wireless/mediatek/mt76/mt76_connac_mcu.c [hunk 5]
```diff
@@ -1552,6 +1552,21 @@ int mt76_connac_mcu_set_deep_sleep(struct mt76_dev *dev, bool enable)
...
+	    new_state == IEEE80211_STA_NOTEXIST)
+		dp = true;
+
+	mt76_connac_mcu_set_deep_sleep(dev, dp);
+
+	return 0;
+}
+EXPORT_SYMBOL_GPL(mt76_connac_sta_state_dp);
+
 void mt76_connac_mcu_coredump_event(struct mt76_dev *dev, struct sk_buff *skb,
```
Comment (LorenzoBianconi) [inline drivers/net/wireless/mediatek/mt76/mt76_connac_mcu.c:1632]: @objelf in order to reduce unnecessary mcu cmds we can do something like:
enable deep sleep switching from IEEE80211_STA_NONE to IEEE80211_STA_NOTEXIST and from IEEE80211_STA_ASSOC to IEEE80211_STA_AUTHORIZED while we can disable it switching from IEEE80211_STA_NOTEXIST to IEEE80211_STA_NONE, what do you think?

File: drivers/net/wireless/mediatek/mt76/mt76_connac_mcu.h (hunk 4 above)
Comment (LorenzoBianconi) [reply drivers/net/wireless/mediatek/mt76/mt76_connac_mcu.h:1114]: @objelf I mean mt76_connac_mcu_roc_event is not upstream yet, but no worries I will rebase it, just send me an updated PR on top of mt7921e-devel branch

File: drivers/net/wireless/mediatek/mt76/mt7921/debugfs.c [hunk 6]
```diff
@@ -236,6 +236,19 @@ mt7921_pm_set(void *data, u64 val)
...
+		mt7921_mutex_release(dev);
+		return 0;
+	}
+	else if(val == 101) {
+		mt7921_mutex_acquire(dev);
+		mt76_connac_mcu_set_deep_sleep(&dev->mt76, 1);
+		mt7921_mutex_release(dev);
+		return 0;
+	}
+
```
Comment (LorenzoBianconi) [inline drivers/net/wireless/mediatek/mt76/mt7921/debugfs.c]: @objelf I would prefer a dedicated debugs knob for enabling/disabling deep sleep avoid magic numbers

File: drivers/net/wireless/mediatek/mt76/mt7921/mac.c [hunk 7]
```diff
@@ -1275,29 +1276,34 @@ mt7921_mac_reset(struct mt7921_dev *dev)
...
+		__mt7921_mcu_drv_pmctrl(dev);
+		clear_bit(MT76_STATE_PM, &dev->mphy.state);
+		dev->pm.stats.last_wake_event = jiffies;
+		dev->pm.stats.last_doze_event = jiffies;
+
 		if (!mt7921_mac_reset(dev))
 			break;
 	}
 	mutex_unlock(&dev->mt76.mutex);
-
```
Comment (LorenzoBianconi) [inline drivers/net/wireless/mediatek/mt76/mt7921/mac.c]: @objelf unnecessary change

File: drivers/net/wireless/mediatek/mt76/mt7921/mac.c [hunk 8]
```diff
@@ -1275,29 +1276,34 @@ mt7921_mac_reset(struct mt7921_dev *dev)
 /* system error recovery */
 void mt7921_mac_reset_work(struct work_struct *work)
 {
-	struct ieee80211_hw *hw;
-	struct mt7921_dev *dev;
+	struct mt7921_dev *dev = container_of(work, struct mt7921_dev, reset_work);
```
Comment (LorenzoBianconi) [inline drivers/net/wireless/mediatek/mt76/mt7921/mac.c]: @objelf please check # of columns here

File: drivers/net/wireless/mediatek/mt76/mt7921/mac.c [hunk 9]
```diff
@@ -1275,29 +1276,34 @@ mt7921_mac_reset(struct mt7921_dev *dev)
...
-	cancel_delayed_work_sync(&dev->pm.ps_work);
-	cancel_work_sync(&dev->pm.wake_work);
 	cancel_work_sync(&dev->phy.roc.work);
 	del_timer_sync(&dev->phy.roc.timer);
 
+	set_bit(MT76_STATE_RESET, &dev->mphy.state);
+	cancel_delayed_work_sync(&pm->ps_work);
+	cancel_work_sync(&pm->wake_work);
+	mt7921_mcu_drv_pmctrl(dev);
+
```
Comment (LorenzoBianconi) [inline drivers/net/wireless/mediatek/mt76/mt7921/mac.c:1324]: @objelf is it enough running __mt7921_mcu_drv_pmctrl() below?

File: drivers/net/wireless/mediatek/mt76/mt76_connac_mac.c [hunk 10]
```diff
@@ -13,6 +13,9 @@ int mt76_connac_pm_wake(struct mt76_phy *phy, struct mt76_connac_pm *pm)
 	if (mt76_is_usb(dev))
 		return 0;
 
+	if (test_bit(MT76_STATE_RESET, &phy->state))
+		return 0;
+
```
Comment (LorenzoBianconi) [inline drivers/net/wireless/mediatek/mt76/mt76_connac_mac.c:18]: @objelf I guess this approach is a bit racy because we can run drv_own or fw_own while running the reset right?

File: drivers/net/wireless/mediatek/mt76/mt76.h [hunk 11]
```diff
@@ -313,6 +313,7 @@ enum {
 	MT76_READING_STATS,
 	MT76_STATE_POWER_OFF,
 	MT76_STATE_SUSPEND,
+	MT76_STATE_RESET,
```
Comment (LorenzoBianconi) [inline drivers/net/wireless/mediatek/mt76/mt76.h]: @objelf we already have MT76_RESET

File: drivers/net/wireless/mediatek/mt76/mt76_connac_mac.c (hunk 10 above)
Comment (objelf) [reply drivers/net/wireless/mediatek/mt76/mt76_connac_mac.c:18]: drv_own and fw_own don't work during wifisys reset

File: drivers/net/wireless/mediatek/mt76/mt76_connac_mcu.h (hunk 4 above)
Comment (objelf) [reply drivers/net/wireless/mediatek/mt76/mt76_connac_mcu.h:1114]: ok, thx

File: drivers/net/wireless/mediatek/mt76/mt76_connac_mcu.c (hunk 5 above)
Comment (objelf) [reply drivers/net/wireless/mediatek/mt76/mt76_connac_mcu.c:1632]: ok, i will try this way and do some stress test.

File: drivers/net/wireless/mediatek/mt76/mt76_connac_mac.c (hunk 10 above)
Comment (LorenzoBianconi) [reply drivers/net/wireless/mediatek/mt76/mt76_connac_mac.c:19]: in the code flowpath we can have both mt7921_reset and fw_own/drv_own running at the same time, right (e.g. on different CPUs). e.g what happen if mt76_connac_pm_wake has already scheduled the wake work?

File: drivers/net/wireless/mediatek/mt76/mt7921/debugfs.c (hunk 6 above)
Comment (objelf) [reply drivers/net/wireless/mediatek/mt76/mt7921/debugfs.c]: ack. i will create dedicated debugs knob for enabling and disabling deep sleep.

File: drivers/net/wireless/mediatek/mt76/mt7921/mac.c [hunk 12]
```diff
@@ -448,16 +448,19 @@ int mt7921_mac_fill_rx(struct mt7921_dev *dev, struct sk_buff *skb)
...
 		status->chain_signal[2] = to_rssi(MT_PRXV_RCPI2, v1);
 		status->chain_signal[3] = to_rssi(MT_PRXV_RCPI3, v1);
-		status->signal = status->chain_signal[0];
-
-		for (i = 1; i < hweight8(mphy->antenna_mask); i++) {
-			if (!(status->chains & BIT(i)))
+		status->signal = -128;
+		for (i = 0; i < hweight8(mphy->antenna_mask); i++) {
+			if (!(status->chains & BIT(i)) ||
+			     (status->chain_signal[i] >= 0))
```
Comment (LorenzoBianconi) [inline drivers/net/wireless/mediatek/mt76/mt7921/mac.c]: @objelf nitpick: no need for () around status->chain_signal[i] >= 0

File: drivers/net/wireless/mediatek/mt76/mt76_connac_mcu.c [hunk 13]
```diff
@@ -837,10 +837,13 @@ int mt76_connac_mcu_add_sta_cmd(struct mt76_phy *phy,
...
 		return PTR_ERR(skb);
 
-	mt76_connac_mcu_sta_basic_tlv(skb, info->vif, info->sta, info->enable);
-	if (info->enable && info->sta)
-		mt76_connac_mcu_sta_tlv(phy, skb, info->sta, info->vif,
-					info->rcpi);
+	if (info->sta) {
+		mt76_connac_mcu_sta_basic_tlv(skb, info->vif, info->sta,
+					      info->enable);
+		if (info->enable && info->sta)
```
Comment (LorenzoBianconi) [inline drivers/net/wireless/mediatek/mt76/mt76_connac_mcu.c]: @objelf we do not need to check info->sta here

File: drivers/net/wireless/mediatek/mt76/mt76_connac_mcu.c [hunk 14]
```diff
@@ -837,10 +837,13 @@ int mt76_connac_mcu_add_sta_cmd(struct mt76_phy *phy,
 	if (IS_ERR(skb))
 		return PTR_ERR(skb);
 
-	mt76_connac_mcu_sta_basic_tlv(skb, info->vif, info->sta, info->enable);
```
Comment (LorenzoBianconi) [inline drivers/net/wireless/mediatek/mt76/mt76_connac_mcu.c:878]: @objelf this is even used in __mt7615_mcu_add_sta(). Can you please check if this change is fine for 7663?

File: drivers/net/wireless/mediatek/mt76/mt76_connac_mac.c (hunk 10 above)
Comment (LorenzoBianconi) [reply drivers/net/wireless/mediatek/mt76/mt76_connac_mac.c:18]: @objelf sorry this comment is related to the change "https://github.com/LorenzoBianconi/wireless-drivers-next/pull/33/commits/0712b8987b3f89947954f9a8819fe890b785c453". Please ignore it

File: drivers/net/wireless/mediatek/mt76/mt76_connac_mcu.c (hunk 14 above)
Comment (objelf) [reply drivers/net/wireless/mediatek/mt76/mt76_connac_mcu.c:878]: Deren mentioned the cmd  usage of mt7921 should be aligned to mt7663

File: drivers/net/wireless/mediatek/mt76/mt76_connac_mcu.c (hunk 13 above)
Comment (objelf) [reply drivers/net/wireless/mediatek/mt76/mt76_connac_mcu.c]: ok, wil remove the redundant check.

**Comment by objelf** [comment]
rebase and update with v2
